Tradução de Async & Performace - Cap. 2 - seção: Nested/Chained Callb… (#347)
cez-aug/You-Dont-Know-JS: async & performance/ch2.md

} );
```

Há uma boa chance de um código como esse ser reconhecível por você. Temos uma cadeia de três funções aninhadas, cada uma representando uma etapa em uma série assíncrona (tarefa, "processo").

Esse tipo de código costuma ser chamado de "callback hell" e às vezes também conhecido como "pirâmide da desgraça" (por sua forma triangular voltada para o lado devido ao recuo aninhado).

> Mas "callback hell" na verdade não tem quase nada a ver com o aninhamento/recuo. É um problema muito mais profundo do que isso. Veremos como e por que à medida que continuarmos no restante deste capítulo.

Reviewer: ```suggestion
Mas "callback hell" na verdade não tem quase nada a ver com o aninhamento/recuo. É um problema muito mais profundo do que isso. Veremos como e porque à medida que continuarmos no restante deste capítulo.
```

Acho que, como estamos falando da justificativa de algo, podemos usar o `porque`; uma alternativa seria `...e o porquê...`.

Author: Faz sentido mesmo.

> Primeiro, aguardamos o evento de "click", depois esperamos o temporizador disparar e, em seguida, aguardamos a resposta da requisação Ajax para processeguir, momento em que pode acontecer tudo novamente.

Reviewer: ```suggestion
Primeiro, aguardamos o evento de "click", depois esperamos o temporizador disparar e, em seguida, aguardamos a resposta da requisação Ajax chegar, momento em que tudo isso pode acontecer novamente.
```
Sobre o `chegar`, não é exatamente a tradução de `to come back` (que seria regressar/voltar), mas o sentido é que esperamos pelo _response_ do _request_; uma alternativa seria tirar e deixar só até `...Ajax,`.

À primeira vista, esse código parece mapear sua assincronia naturalmente para o planejamento sequencial do cérebro.

Primeiro (*agora*), nós:

```js
listen( "..", function handler(..){
	// ..
} );
```

> Depois *mais tarde*, nós:

Reviewer: ```suggestion
Depois, *mais tarde*, nós:
```

```js
setTimeout( function request(..){
	// ..
}, 500) ;
```

> Depois ainda *mais tarde*, nós:

Reviewer: ```suggestion
Depois, ainda *mais tarde*, nós:
```

else ..
```

Mas há vários problemas em raciocinar sobre esse código linearmente dessa maneira.

> Primeiramente, é um equívoco do exemplo que nossos passos estão nas linhas subsequentes (1, 2, 3 e 4...). Em programas JS assíncronos reais, geralmente há muito mais ruído bagunçando as coisas, ruído que temos que manobrar habilmente em nossos cérebros conforme saltamos de uma função para a próxima. Compreender o fluxo assíncrono em tal código carregado de callbacks não é impossível, mas certamente não é natural ou fácil, mesmo com muita prática.

Reviewer: ```suggestion
Primeiramente, é um equívoco do exemplo que nossos passos estejam em linhas subsequentes (1, 2, 3 e 4...). Em programas JS assíncronos reais, geralmente há muito mais ruído bagunçando as coisas, ruído que temos que manobrar habilmente em nossos cérebros conforme saltamos de uma função para a próxima. Compreender o fluxo assíncrono em tal código carregado de callbacks não é impossível, mas certamente não é natural ou fácil, mesmo com muita prática.
```

doF();
```

> Embora os mais experientes identificarão a verdadeira ordem das operações aqui, eu aposto que ela é mais do que uma pequena confusão à primeira vista e custa alguns ciclos mentais conbinados para chegar na resposta certa. As operações acontecerão nesta ordem:

Reviewer: ```suggestion
Embora os mais experientes identificarão a verdadeira ordem das operações aqui, eu aposto que ela é mais do que uma pequena confusão à primeira vista e custa alguns ciclos mentais combinados para chegar na resposta certa. As operações acontecerão nesta ordem:
```

* `doE()`
* `doD()`

Você acertou na primeira vez que olhou o código?

> OK, alguns de vocês estão pensando que fui injusto na nomeação das minhas funções, para intencionalmente desviá-los. Juro que estava apenas nomeando na ordem em que aparecem, de cima para baixo. Mas deixe-me tentar novamente:

Reviewer: ```suggestion
OK, alguns de vocês estão pensando que fui injusto na nomeação das minhas funções, para fazer vocês se perderem intencionalmente. Juro que estava apenas nomeando na ordem em que aparecem, de cima para baixo. Mas deixe-me tentar novamente:
```

Author: Ficou ótima sua sugestão! :clap:

doB();
```

> Agora, eu as nomeei alfabeticamente em ordem de execução real. Mas eu ainda aposto, que mesmo agora com experiência nesse cenário, ordenar as operações na ordem `A -> B -> C -> D -> E -> F` não é natural para muitos ou nenhum de vocês leitores. Certamente seus olhos saltam muito para cima e para baixo no trecho de código, certo?

Reviewer: ```suggestion
Agora, eu as nomeei alfabeticamente em ordem de execução real. Mas eu ainda aposto, que mesmo agora com experiência nesse cenário, ordenar as operações na ordem `A -> B -> C -> D -> E -> F` não é natural para muitos ou nenhum de vocês leitores. Certamente seus olhos saltaram várias vezes para cima e para baixo no trecho de código, certo?
```

Mas mesmo que tudo isso seja natural para você, há ainda mais um perigo que pode causar estragos. Você consegue identificar qual é?

E se `doA(..)` ou `doD(..)` não forem realmente assíncronas, da maneira que obviamente assumimos que fossem? Oh oh, agora a ordem é diferente. Se ambas forem síncronas (e talvez apenas algumas vezes, dependendo das condições do programa no momento), a ordem agora é `A -> C -> D -> F -> E -> B`.

> Esse som que você acabou de ouvir fracamente ao fundo são os suspiros de milhares de desenvolvedores JS que tiveram um momento de frustração.

Reviewer: ```suggestion
Esse som que você acabou de ouvir levemente ao fundo são os suspiros de milhares de desenvolvedores JS que tiveram um momento de frustração.
```

O aninhamento é o problema? É isso que torna tão difícil rastrear o fluxo assíncrono? Isso faz parte, certamente.

> Mas deixe-me reescrever o exemplo anterior com evento/tempo limite/Ajax aninhados sem usar aninhamento:

Reviewer: ```suggestion
Mas deixe-me reescrever o exemplo anterior com evento/timeout/Ajax aninhados sem usar aninhamento:
```

"_timeout_" é um termo que faz parte da linguagem, não acho uma boa traduzir.
De qualquer forma, se achar que é melhor manter a tradução, mesmo que não seja literal, "temporizador" encaixa melhor nesse contexto.

}
```

> Essa formulação do código não é tão identificável quanto aos problemas de aninhamento/recuo de sua forma anterior e, ainda assim, é tão suscetível ao "callback hell". Por quê?

Reviewer: ```suggestion
Essa formulação do código não é tão identificável quanto aos problemas de aninhamento/indentação de sua forma anterior e, ainda assim, é tão suscetível ao "callback hell". Por quê?
```

À medida que raciocinamos linearmente (sequencialmente) sobre esse código, temos que pular de uma função para a próxima e para a próxima, saltando por toda a base do código para "ver" o fluxo da sequência. E lembre-se, esse é um código simplificado na melhor das hipóteses. Todos nós sabemos que as bases de código de programas JS assíncronos reais costumam ser fantasticamente mais confusas, o que torna essas ordens de magnitude de raciocínio mais difíceis.

> Outra coisa a observar: para vincular as etapas 2, 3 e 4 a fim de que ocorram em sucessão, a única possibilidade que os callbacks nos fornece é codificar a etapa 2 na etapa 1, a etapa 3 na etapa 2, a etapa 4 na etapa 3, e assim por diante. A codificação não é necessariamente uma coisa ruim, se realmente for uma condição fixa que a etapa 2 sempre deve levar à etapa 3.

Reviewer: ```suggestion
Outra coisa a observar: para vincular as etapas 2, 3 e 4 a fim de que ocorram em sucessão, a única possibilidade que os callbacks nos fornecem é codificar a etapa 2 na etapa 1, a etapa 3 na etapa 2, a etapa 4 na etapa 3, e assim por diante. A codificação não é necessariamente uma coisa ruim, se realmente for uma condição fixa que a etapa 2 sempre deve levar à etapa 3.
```
Sobre as traduções dos termos `hardcode` e `hardcoding`, o que acha de tentar uma ideia de _definir explicitamente/fixamente_?

Author: Acredito que "_definir fixamente_" fica um pouco melhor realmente. Traduzimos assim então?

Reviewer: Sim, por favor!

> Mas forçar a codificação dessa maneira definitivamente torna o código um pouco mais frágil, pois não leva em consideração nada de errado que possa causar um desvio na progressão das etapas. Por exemplo, se a etapa 2 falhar, a etapa 3 nunca será alcançada, nem a etapa 2 tentará novamente ou moverá para um fluxo alternativo de tratamento de erros e assim por diante.

Reviewer: O que acha?
```suggestion
Mas forçar o código dessa maneira definitivamente torna o código um pouco mais frágil, pois não leva em consideração nada de errado que possa causar um desvio na progressão das etapas. Por exemplo, se a etapa 2 falhar, a etapa 3 nunca será alcançada, nem a etapa 2 tentará novamente ou moverá para um fluxo alternativo de tratamento de erros e assim por diante.
```

Todos esses problemas são coisas que você *pode* codificar manualmente em cada etapa, mas esse código costuma ser muito repetitivo e não pode ser reutilizado em outras etapas ou em outros fluxos assíncronos em seu programa.

Mesmo que nossos cérebros possam planejar uma série de tarefas de uma forma sequencial (isso, depois isso, então isso), a natureza de eventos da nossa operação cerebral torna a recuperação/retentativa/bifurcação do controle de fluxo quase sem esforço. Se você está fazendo compras e percebe que deixou a lista em casa, o dia não termina porque você não planejou isso com antecedência. Seu cérebro contorna esse contratempo facilmente: você vai para casa, pega a lista e volta direto para a loja.

Mas a natureza frágil de callbacks codificados manualmente (mesmo com tratamento de erros codificados permanentemente) é com frequência muito menos elegante. Depois que você acaba especificando (também conhecido como pré-planejamento) todas as várias eventualidades/caminhos, o código se torna tão complicado que é difícil mantê-lo ou atualizá-lo.

> **Isto** é o que se trata o "callback hell"! O aninhamento/recuo são basicamente um espetáculo à parte, uma pista falsa.

Reviewer: ```suggestion
**Isto** é o que se trata o "callback hell"! O aninhamento/indentação são basicamente um espetáculo à parte, uma pista falsa.
```

Author: Sobre o "That", não seria melhor traduzi-lo para "Isso" por se tratar de algo que já foi abordado?

Reviewer: Acho que encaixa sim.

E como se isso não bastasse, ainda nem tocamos no que acontece quando duas ou mais cadeias dessas continuações de callback estão acontecendo *simultaneamente*, ou quando a terceira etapa se ramifica em callbacks "paralelos" com portas ou travas, ou... oh céus!, meu cérebro dói, e o seu!?

> Você está entendendo a ideia aqui de que nosso cérebro, que se comporta de maneira sequencial e bloqueante, simplesmente não mapeiam bem código assíncrono orientado a callback? Esta é a primeira grande deficiência a ressaltar sobre callbacks: eles expressam assincronia em código de maneira que nosso cérebro trabalha apenas para mantê-lo de forma síncrona (trocadilho intencional!).

Reviewer: ```suggestion
Você está entendendo a ideia aqui de que nosso cérebro, que se comporta de maneira sequencial e bloqueante, simplesmente não mapeiam bem código assíncrono orientado a callback? Esta é a primeira grande deficiência a ressaltar sobre callbacks: eles expressam assincronia em código de maneira que nosso cérebro luta apenas para mantê-lo de forma síncrona (trocadilho intencional!).
```
